From 82a2d5d6bf726eabfde31ff93c1b917b04ccf8cb Mon Sep 17 00:00:00 2001
From: "Chew, Tong Liang" <tong.liang.chew@intel.com>
Date: Fri, 27 Nov 2020 01:53:21 +0800
Subject: [PATCH] Adding dummy modules_install

Signed-off-by: Chew, Tong Liang <tong.liang.chew@intel.com>
---
 Makefile | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/Makefile b/Makefile
index d595cc4..a243a73 100644
--- a/Makefile
+++ b/Makefile
@@ -14,3 +14,5 @@ test:	all
 	grep -q '^kvmfr' /proc/modules && sudo rmmod kvmfr || true
 	sudo insmod ./kvmfr.ko
 	sudo chown $(USER) /dev/uio0
+
+modules_install:
-- 
2.27.0

